src/main/java/pro/belbix/ethparser/web3/harvest/HardWorkCalculator.java

```diff
@@ -0,0 +1,119 @@
+import pro.belbix.ethparser.repositories.HarvestRepository;
+import pro.belbix.ethparser.web3.EthBlockService;
+import pro.belbix.ethparser.web3.prices.PriceProvider;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class HardWorkCalculator {
+
+    private final double ETH_ESTIMATE = 0.1;
+    private final PriceProvider priceProvider;
+    private final HarvestRepository harvestRepository;
+    private final HardWorkRepository hardworkRepository;
+    private final EthBlockService ethBlockService;
+
+    public HardWorkCalculator(PriceProvider priceProvider, HarvestRepository harvestRepository, HardWorkRepository hardWorkRepository, EthBlockService ethBlockService) {
+        this.priceProvider = priceProvider;
+        this.harvestRepository = harvestRepository;
+        this.hardworkRepository = hardWorkRepository;
+        this.ethBlockService = ethBlockService;
+    }
+
+    public double calculateTotalHardWorksFeeByOwner(String ownerAddress) {
+        long lastBlock = ethBlockService.getLastBlock();
+        long lastBlockDate = ethBlockService.getTimestampSecForBlock(null, lastBlock);
+        
+        List<HarvestDTO> harvests = harvestRepository.fetchAllByOwner(ownerAddress, 0, lastBlockDate);
+        List<HarvestDTO> sortedHarvests = harvests
+            .stream()
+            .sorted(Comparator.comparingLong(HarvestDTO::getBlockDate))
+            .collect(Collectors.toList());
+        
+        HashMap<String, ArrayList<ArrayList<Long>>> blockRangeByVault = new HashMap<>();
+        for (HarvestDTO harvest : sortedHarvests) {
+            String vault = harvest.getVault();
+            double balance = harvest.getOwnerBalance();
+
+            if (!blockRangeByVault.containsKey(vault)) {
+                if (balance == 0) {
+                    continue;
+                }
+                ArrayList<ArrayList<Long>> blockPeriods = new ArrayList<>();
+                blockRangeByVault.put(vault, blockPeriods);
+            }
+
+            updateBlockPeriods(blockRangeByVault.get(vault), harvest);
+        }
+
+        return blockRangeByVault
+            .entrySet()
+            .stream()
+            .map(entry -> {
+                String vault = entry.getKey();
+
+                ArrayList<ArrayList<Long>> blockPeriods = entry.getValue();
+                return blockPeriods
+                    .stream()
+                    .map(blockPeriod -> {
+                        Long blockStartDate = blockPeriod.get(0);
+                        Long blockEndDate = lastBlockDate;
+                        if (blockPeriod.size() > 1) {
+                            blockEndDate = blockPeriod.get(1);
+                        }
+
+                        return hardworkRepository.findAllByVaultOrderByBlockDate(vault, blockStartDate, blockEndDate);
```

Comment:
it will generate a ton of db queries
need to reduce it and use only one
you can get all hardWorks for used vaults for minBlock-maxBlock and handle this array

Reply:
Great spot on this!

I think there a balance between the API's memory cost and the DB's lookup cost. If we fetch all the hardworks for all related vaults with the widest span of the periods in one go, it would cost more in terms of memory, and there could be a lot of unnecessary records(the ones that are not within the effective periods) loaded that needs to be filtered out in-memory and not being able to make use of the DB indexes.

But as you pointed out, it could end up a lot of SQL queries, which in certain setup could be slow especially due to the network latency between the API and the DB instance.

For now, I have updated the DB look-up to be per vault with min-block and max-block. Thus, the performance tradeoff is somewhere between your suggestion and my initial implementation.  😄